File: go/todo/todo/parser.go
package todo

import (
	"strings"
	"time"
)

type Parser struct {
}

func (parser *Parser) Parse(text string) []Item {
	items := []Item{}

	lines := strings.Split(text, "\n")

	for _, line := range lines {
		first := ""
		if len(line) > 0 {
			first = string(line[0])
		}
		switch first {
		case "#":
			items = append(items, NewComment(line[1:len(line)]))
		case "*":
			task_string := line[1:len(line)]
			parts := strings.Split(task_string, "-!-")
			task := Task{identifier: strings.Trim(parts[0], " -!- "), estimated_time: strings.Trim(parts[1], " -!- "), priority: strings.Trim(parts[2], " -!- ")}
			items = append(items, task)
		case "-":
			task_string := line[1:len(line)]
			parts := strings.Split(task_string, "-!-")
			completed_at, _ := time.Parse(DATE_FORMAT, parts[3])
			task := Task{identifier: strings.Trim(parts[0], " -!- "), estimated_time: strings.Trim(parts[1], " -!- "), priority: strings.Trim(parts[2], " -!- "), completed_at: completed_at}
			items = append(items, task)
		default:
			items = append(items, NewSpacer())
		}
	}

	return items
}

File: java/src/main/java/me/ramblingsby/todos/Parser.java
package me.ramblingsby.todos;

import java.util.ArrayList;
import java.util.List;

public class Parser {
  public List<Item> parse(String string) {
    List<Item> items = new ArrayList<Item>();

    String[] lines = string.split("\n");

    for (String line : lines) {
      String text;
      String[] parts;

      try {
        switch (line.charAt(0)) {
          case '#':
            text = line.substring(1);
            items.add(new Comment(text));
          break;
          case '*':
            text = line.substring(1);
            parts = text.split("-!-");
            if (parts.length == 3) {
              items.add(new Task(parts[0].trim(), parts[1].trim(), parts[2].trim()));
            } else if (parts.length == 2) {
              items.add(new Task(parts[0].trim(), parts[1].trim(), ""));
            } else if (parts.length == 1) {
              items.add(new Task(parts[0].trim(), "", ""));
            }
          break;
          case '-':
            text = line.substring(1);
            parts = text.split("-!-");
            items.add(new Task(parts[0].trim(), parts[1].trim(), parts[2].trim(), parts[3].trim()));
          break;
          default:
            items.add(new Spacer());
        }
      } catch (IndexOutOfBoundsException e) {
        items.add(new Spacer());
      }
    }

    return items;
  }
}
